staging/src/k8s.io/apimachinery/pkg/conversion/converter.go

```diff
@@ -442,8 +442,33 @@ func (c *Converter) doConversion(src, dest interface{}, flags FieldMatchingFlags
 	if fn, ok := c.generatedConversionFuncs.untyped[pair]; ok {
 		return fn(src, dest, scope)
 	}
-	// TODO: consider everything past this point deprecated - we want to support only point to point top level
-	// conversions
+
+	if pair.source == pair.dest {
+		if _, err := EnforcePtr(src); err != nil {
+			return err
+		}
+		if _, err := EnforcePtr(dest); err != nil {
+			return err
+		}
+		// TODO: Ideally, to avoid this hack, we should auto-generate self-conversions.
+		// For now, we assume that having self-conversions is needed only for top-level
+		// API objects, for which we autogenerated DeepCopyInto() method that we use.
+		deepCopyIntoName := "DeepCopyInto"
+		if _, ok := pair.source.MethodByName(deepCopyIntoName); ok {
+			deepCopyIntoType := reflect.ValueOf(src).MethodByName(deepCopyIntoName).Type()
+			if deepCopyIntoType.NumIn() == 1 && deepCopyIntoType.In(0) == pair.dest && deepCopyIntoType.NumOut() == 0 {
+				reflect.ValueOf(src).MethodByName(deepCopyIntoName).Call([]reflect.Value{reflect.ValueOf(dest)})
+				// TODO: Conversions were cleaning up TypeMeta field, which DeepCopy is copying.
+				//   It seems, it was causing only one unit test to failed, so for now we don't clean it up.
+				return nil
+			}
+		}
+	}
+
+	return fmt.Errorf("converting (%s) to (%s): unknown conversion", pair.source, pair.dest)
```

Comment:
nit: trim the `*` prefix from pair.source.String() and pair.dest.String() to keep the error messages identical?

Reply:
Done (though in slightly different way, by keeping the preexisting EnforcePtr stuff).